### PR TITLE
feat(gatsby-image): Responsive Art Direction

### PR DESCRIPTION
**Pull request by polarathene**
## Description

Presently, a render is triggered from an `onload` event calling `handleImageLoaded()` and fluid type `aspectRatio` or fixed type `width` & `height` state values for CSS are only updated to the new variant after it's loaded.

Placeholder lazy loading state is also not reset as the component isn't aware of when loading a new variant begins. If the image variant is cached, there is still a delay with the `onload` event until the image request has returned a response from the server; a noticeable delay on high latency connections - breaking layout (as the content/component has not been updated yet to fit the new breakpoint).

`matchMedia()` resolves that. We can then detect if the image has been cached internally and then trigger a render which might leverage browser cache detection, otherwise reset the lazy load placeholder behavior.

See individual commits for more logical grouping and further details related to those additions.

May need to rebase after the following PRs are merged:
- [x] Dependent on https://github.com/gatsbyjs/gatsby/pull/26118 to pass tests.
- [x] Implements a related conditional as [this PR](https://github.com/gatsbyjs/gatsby/pull/26117).

My "deprecated" comment for [`MediaQueryListListener`](https://developer.mozilla.org/en-US/docs/Web/API/MediaQueryListListener)(note, modern browsers alias this for now) is not quite correct, [MDN encourages using the newer standard event mechanism](https://developer.mozilla.org/en-US/docs/Web/API/MediaQueryList/addListener), the older non-standard API is kept by browsers for backward compatibility. A newer version of Safari that's to be released this year will also [support the newer API](https://github.com/microsoft/TypeScript/issues/32210#issuecomment-652745256). [MDN browser compatibility](https://developer.mozilla.org/en-US/docs/Web/API/MediaQueryList#Browser_compatibility)

## Related Issues

Originally implemented [in this PR](https://github.com/gatsbyjs/gatsby/pull/24811#issuecomment-657387494), which has been split out into several smaller ones.

### CODE REVIEW

**Comment by pvdz** [comment]
Dependent PR has been merged, ready to rebase this.

Headsup that I probably won't get around to take a look until later today or tomorrow.

Additionally, I'd expect @wardpeet to also hope for e2e coverage for this PR. Especially if none of the tests need to change for that (but perhaps you're waiting to rebase before changing those).

**Comment by polarathene** [comment]
> hope for e2e coverage for this PR. 

I'd need some guidance on how to approach that. 

I don't think there is currently any high-latency network tests or any tests that track CSS transition or caching behaviour? They're generally assuming that if X and Y inputs are provided, Z output is expected, these kind of tests would be more nuanced..

> Especially if none of the tests need to change for that (but perhaps you're waiting to rebase before changing those).

The tests only broke and required the proper Jest `matchMedia()` mock due to not being able to run the new code using other `matchMedia()` API surface in Jest's JSDOM (which lacks any support for the API).

The existing Art Direction tests ensure the image source changes accordingly to meet it's media condition, but that's it.

I don't expect the rebase to raise any errors, but we'll see shortly.

File: packages/gatsby-image/src/index.js [hunk 1]
```diff
@@ -362,6 +362,31 @@ class Image extends React.Component {
...
+        this.mq = { base }
+        this.mq.queries = imageVariants.filter(v => v.media)
+        this.mq.queries.forEach(v => {
+          this.mq[v.media] = v
+          const mql = window.matchMedia(v.media)
+          if (mql.addEventListener) {
+            mql.addEventListener(`change`, this.handleVariantChange)
+          } else {
+            // Deprecated 'MediaQueryList' API, <Safari 14, IE, <Edge 16
+            mql.addListener(this.handleVariantChange)
```
Comment (wardpeet) [inline packages/gatsby-image/src/index.js:385]: We probably want to clean it up with `componentWillUnmount`. Besides that this looks great! Thank you!

Comment (wardpeet) [reply packages/gatsby-image/src/index.js:385]: I'm happy to take over as you've been waiting long on this feedback.

**Comment by polarathene** [comment]
Rebased onto master.

@wardpeet good catch there, sorry about missing that! 

Handing this PR over to you to wrap up thanks.

**Comment by wardpeet** [comment]
@polarathene Thank you, I'm grateful that you're taking the time to finish it. Awesome work so far! 👍 👏

**Comment by polarathene** [comment]
> I'm grateful that you're taking the time to finish it.

Sorry, we have a misunderstanding. You said earlier you'd be happy to take over the PR, and I thank you for that. I need to prioritize my time on other tasks that get me closer towards earning a living and landing my first job as a web developer.

I've only been handling the other related PR that doesn't seem like it'll be merged, as expecting users to spend time investigating `gatsby-image` internals and each DIY an Art Direction fix with a HOC isn't a pleasant experience to offload to the user (I should know, considering how much time I have sunk into resolving these issues :P ). Since I know `gatsby-image` well enough, I updated that PR to at least give other users a good reference that should work well.

**Comment by wardpeet** [comment]
OH yeah my bad "Handing this PR over to you to wrap up thanks." I'll take it over :) 

> I've only been handling the other related PR that doesn't seem like it'll be merged, as expecting users to spend time investigating gatsby-image internals and each DIY an Art Direction fix with a HOC isn't a pleasant experience to offload to the user (I should know, considering how much time I have sunk into resolving these issues :P ). Since I know gatsby-image well enough, I updated that PR to at least give other users a good reference that should work well.

I know, that's why we are in desperate need to create a new gatsby-image that properly tell the user what it can do and what you need to do yourself.

**Comment by polarathene** [comment]
> I'll take it over :)

Thanks!

> I know, that's why we are in desperate need to create a new gatsby-image

Wasn't that meant to be solved with the refactor to hooks? I had inquired about that to you in an issue/PR back around May-June, asking if it was being worked on publicly or if the core team had a private refactoring going on. Offered to collaborate on that, but never heard back. 

I'm not currently in a position to have spare time for a hooks refactor anymore, but if my situation improves within the next 6 months, I might be able to give it a shot if there's still no progress. Originally I wanted to get these June PRs discussed with the team, iterated on and merged, then tackle the hooks refactor in advance for Gatsby consider when ready.

**Comment by wardpeet** [comment]
I have something to share in a few days :) Here is a sneak peak, I've only started on it a few weeks ago and mostly was exploration on how to fix LCP web-vitals.

```
    <React.Fragment>
      <div>Hello world!</div>
      <GatsbyImage
        width={500}
        height={600}
        placeholder={{
          fallback: fixedLong.base64,
        }}
        images={{
          fallback: {
            src: fixedLong.src,
            srcSet: fixedLong.srcSet,
          },
          sources: fixedImages,
        }}
        alt="hello"
      />
      <GatsbyImage
        width={500}
        height={600}
        loading="eager"
        placeholder={{
          fallback: fixedLongEager.base64,
        }}
        images={{
          fallback: {
            src: fixedLongEager.src,
            srcSet: fixedLongEager.srcSet,
          },
          sources: fixedImages,
        }}
        alt="hello"
      />

      <GatsbyImageHydrator>
        some extra stuff
        <div {...getWrapperProps()}>
          <Placeholder {...getPlaceholderProps()} />
          <MainImage {...getMainImageProps()} alt="zzz" />
        </div>
      </GatsbyImageHydrator>
    </React.Fragment>
```

**Comment by polarathene** [comment]
Minor update to accommodate the request [here](https://github.com/gatsbyjs/gatsby/pull/26097#issuecomment-694116108) to avoid `isHydrated` adding an unnecessary render for anyone not using Art Direction.

**Comment by wardpeet** [comment]
Closing, this is fixed in our new [gatsby-plugin-image](https://github.com/gatsbyjs/gatsby/tree/master/packages/gatsby-plugin-image) component. Thank you for your contribution! <3

**Pull request closed without merging**